edge-modules/MetricsCollector/MetricsScrapeAndUpload.cs

```diff
@@ -38,18 +41,44 @@ public void Dispose()
 
         async Task ScrapeAndUploadMetricsAsync(CancellationToken cancellationToken)
         {
+            IEnumerable<Metric> scrapedMetrics = Enumerable.Empty<Metric>();
             try
             {
-                IEnumerable<Metric> metrics = await this.scraper.ScrapeEndpointsAsync(cancellationToken);
-                await this.publisher.PublishAsync(this.GetGuidTaggedMetrics(metrics), cancellationToken);
+                await this.PublishPreviouslyFailedMetricsAsync(cancellationToken);
+
+                scrapedMetrics = await this.scraper.ScrapeEndpointsAsync(cancellationToken);
+                scrapedMetrics = this.GetGuidTaggedMetrics(scrapedMetrics);
+                await this.publisher.PublishAsync(scrapedMetrics, cancellationToken);
                 Logger.LogInformation("Successfully scraped and uploaded metrics");
             }
             catch (Exception e)
             {
                 Logger.LogError(e, "Error scraping and uploading metrics");
+                this.failedUploadMetrics.AddRange(scrapedMetrics);
             }
         }
 
+        async Task PublishPreviouslyFailedMetricsAsync(CancellationToken cancellationToken)
+        {
+            int processedCount = 0;
+            int batchSize = 200;
+            foreach (IEnumerable<Metric> batch in this.failedUploadMetrics.Batch(batchSize))
+            {
+                try
+                {
+                    await this.publisher.PublishAsync(batch, cancellationToken);
+                    processedCount += batchSize;
+                }
+                catch (Exception e)
+                {
+                    Logger.LogError(e, "Error uploading metrics from prior scrape");
+                    break;
+                }
+            }
+
+            this.failedUploadMetrics.RemoveRange(0, Math.Min(processedCount, this.failedUploadMetrics.Count));
+        }
+
         IEnumerable<Metric> GetGuidTaggedMetrics(IEnumerable<Metric> metrics)
```

Comment:
Please rename "ApplyAdditionalMetricTags"; or suggest to change it as extension method and name it as "WithAdditionalMetricTags".